tracing-attributes/src/lib.rs

```diff
@@ -262,14 +273,40 @@ pub fn instrument(args: TokenStream, item: TokenStream) -> TokenStream {
     // If the function is an `async fn`, this will wrap it in an async block,
     // which is `instrument`ed using `tracing-futures`. Otherwise, this will
     // enter the span and then perform the rest of the body.
+    // If `err` is in args, instrument any resulting `Err`s.
     let body = if asyncness.is_some() {
-        quote_spanned! {block.span()=>
-            tracing_futures::Instrument::instrument(
-                async move { #block },
-                __tracing_attr_span
-            )
-                .await
+        if instrument_err(&args) {
+            quote_spanned! {block.span()=>
+                tracing_futures::Instrument::instrument(async move {
+                    match async move { #block }.await {
+                        Ok(x) => Ok(x),
+                        Err(e) => {
+                            tracing::error!("{}", e);
+                            Err(e)
+                        }
+                    }
+                }, __tracing_attr_span).await
+            }
+        } else {
+            quote_spanned! {block.span()=>
+                tracing_futures::Instrument::instrument(
+                    async move { #block },
+                    __tracing_attr_span
+                )
+                    .await
+            }
         }
+    } else if instrument_err(&args) {
+        quote_spanned!(block.span()=>
+            let __tracing_attr_guard = __tracing_attr_span.enter();
+            match { #block } {
+                Ok(x) => Ok(x),
+                Err(e) => {
+                    tracing::error!("{}", e);
+                    Err(e)
+                }
+            }
+        )
```

Comment:
It would be nice to have a way to reduce duplication between the async and non-async versions of the generated code. Not sure if that's practical, though...

Reply:
I tried splitting it out into two separate conditional blocks (redefining `block` first to add the match arm, then leaving the existing `if asyncness.is_some()` code), but couldn't get it to work.

Reply:
Okay, no worries! I'm fine with this as-is, I was just wondering if you'd tried that. Thanks!